specutils/spectra/spectrum_collection.py

```diff
@@ -0,0 +1,180 @@
+                end_bin = max([x.spectral_axis[-1] for x in self._items]).value
+
+                if grid == 'coarse':
+                    largest_bin = max([np.max(np.diff(x.spectral_axis))
+                                    for x in self._items])
+                    resample_grid = np.arange(
+                        start_bin, end_bin, largest_bin.value) * largest_bin.unit
+                elif grid == 'fine':
+                    smallest_bin = min([np.min(np.diff(x.spectral_axis))
+                                        for x in self._items])
+                    resample_grid = np.arange(
+                        start_bin, end_bin, smallest_bin.value) * smallest_bin.unit
+            else:
+                resample_grid = None
+
+        # Ensure the unit information of the resample grid
+        resample_grid = u.Quantity(resample_grid,
+                                   self._items[0].spectral_axis_unit)
+
+        if resample_grid is not None:
+            resampled_spectra = [x.resample(resample_grid) for x in self._items]
+        else:
+            resampled_spectra = None
+
+        # Next, resample the spectrum objects to this grid
+        return resample_grid, resampled_spectra
+
+
+class SpectrumCollection(SpectrumArray, ResampleMixin, NDIOMixin):
+    """
+    A container class for spectra which themselves do not share a single
+    dispersion solution. :class:`~specutils.Spectrum1D` objects added to this
+    collection are automatically rebinned onto the user-specified dispersion
+    grid.
+
+    Parameters
+    ----------
+    output_grid : str, array-like, tuple
+        See the docstring in
+        :class:`~specutils.spectra.spectrum_collection.ResampleMixin`.
+    """
+    def __init__(self, items=[], output_grid=None):
+        super(SpectrumCollection, self).__init__(items)
+
+        self._output_grid = output_grid if output_grid is not None else 'coarse'
+        self._resampled_grid, self._resampled_items = None, None
+        self._evaluate_grids()
+
+    def __setitem__(self, *args, **kwargs):
+        super(SpectrumCollection, self).__setitem__(*args, **kwargs)
+        self._evaluate_grids()
+
+    def __delitem__(self, *args, **kwargs):
+        super(SpectrumCollection, self).__delitem__(*args, **kwargs)
+        self._evaluate_grids()
+
+    def insert(self, *args, **kwargs):
+        super(SpectrumCollection, self).insert(*args, **kwargs)
+        self._evaluate_grids()
+
+    def append(self, *args, **kwargs):
+        super(SpectrumCollection, self).append(*args, **kwargs)
+        self._evaluate_grids()
+
+    @property
+    def output_grid(self):
+        return self._output_grid
+
+    @output_grid.setter
+    def output_grid(self, value):
+        """
+        If the output grid value is changed, force the `SpectrumCollection` to
+        re-evaluate the resampled grid.
+        """
+        self._output_grid = value
+        self._evaluate_grids()
+
+    def _evaluate_grids(self):
```

Comment:
This seems potentially expensive.  I haven't thought through it, but can this be cached...?

Reply:
`_evaluate_grids` should only be called in a few cases:

1. Then the `SpectrumCollection` is created
2. When an item has been added, removed, or replaced in the `SpectrumCollection`.

Those seem like reasonable times to re-evaluate the grid?

Reply:
Yes.  It shows up in many lines of code, so I wasn't sure if there were more cases.